ipn/ipnlocal/state_test.go

```diff
@@ -0,0 +1,778 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package ipnlocal
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"tailscale.com/control/controlclient"
+	"tailscale.com/ipn"
+	"tailscale.com/tailcfg"
+	"tailscale.com/types/empty"
+	"tailscale.com/types/logger"
+	"tailscale.com/types/netmap"
+	"tailscale.com/wgengine"
+)
+
+// notifyThrottler receives notifications from an ipn.Backend, blocking
+// if they were unexpected and complaining if they are too few.
```

Comment:
The implementation seems to crash (t.Fatal) on unexpected counts, rather than block. Docstring needs adjusting?

Reply:
I'll add more to the docstring, but the idea is that if there is a goroutine generating a sequence of events in the right order in the background, it will block delivering notifications if it tries to go too fast. That keeps the test and the main program in lockstep, which is important so that we can generate fake controlclient events at exactly the right moment. (This feature isn't used very heavily though. Mostly if it blocks, that's just a sign the test is going to fail, so we time out and it does log.Fatalf.)